bottlerocket/agents/src/bin/ecs-workload-agent/main.rs

```diff
@@ -0,0 +1,499 @@
+        .context(error::NoTaskSnafu)?;
+    let running_count = tasks
+        .iter()
+        .filter(|task| task.last_status() == Some("STOPPED"))
+        .filter(|task| task.stop_code() == Some(&TaskStopCode::EssentialContainerExited))
+        .filter(|task| {
+            task.containers()
+                .unwrap_or_default()
+                .iter()
+                .filter(|container| container.exit_code() != Some(0))
+                .count()
+                == 0
+        })
+        .count() as i32;
+    Ok(TestResults {
+        outcome: if 1 == running_count {
+            Outcome::Pass
+        } else {
+            Outcome::Fail
+        },
+        num_passed: running_count as u64,
+        num_failed: (1 - running_count) as u64,
+        num_skipped: 0,
+        other_info: None,
+    })
+}
+
+/// Waits for the cluster to have registered instances.
+async fn wait_for_cluster_ready(
+    ecs_client: &aws_sdk_ecs::Client,
+    cluster: &str,
+) -> Result<(), Error> {
+    loop {
+        let cluster = ecs_client
+            .describe_clusters()
+            .clusters(cluster)
+            .send()
+            .await
+            .context(error::ClusterDescribeSnafu)?
+            .clusters()
+            .context(error::NoTaskSnafu)?
+            .first()
+            .context(error::NoTaskSnafu)?
+            .clone();
+
+        if cluster.registered_container_instances_count() != 0 {
+            return Ok(());
+        }
+        tokio::time::sleep(Duration::from_secs(2)).await;
+    }
+}
+
+/// Retrieves the task_definition and revision of a matching test task definition. If the
+/// task definition doesn't exist, it will be created. If the task definition exists but
+/// the container defintions doesn't match the current settings in the config, a new task
+/// definition revision will be created.
+async fn create_or_find_task_definition(
+    ecs_client: &aws_sdk_ecs::Client,
+    test_def: &WorkloadTest,
+) -> Result<String, Error> {
+    let task_def_name = format!("{}{}", DEFAULT_TASK_DEFINITION_PREFIX, test_def.name);
+    let task_info = ecs_client
+        .describe_task_definition()
+        .task_definition(task_def_name.clone())
+        .send()
+        .await;
+
+    if let Err(SdkError::ServiceError { err, raw: _ }) = &task_info {
+        // If we get an error and it's a ClientException, that means the call worked
+        // but the identifier was not found. Anything else and it will be handled
+        // below with lookup calls.
+        if matches!(
+            &err.kind,
+            DescribeTaskDefinitionErrorKind::ClientException(_)
+        ) {
+            return create_task_definition(ecs_client, &task_def_name, test_def).await;
+        }
+    }
+
+    // There is a task definition with the right name, now we need to find the right
+    // revision that matches our image:tag.
+    if let task_arn @ Ok(_) = find_task_rev(ecs_client, &task_def_name, test_def).await {
```

Comment:
```suggestion
    if let Some(task_arn) = find_task_rev(ecs_client, &task_def_name, test_def).await.ok() {
```

Reply:
This isn't correct. We want to get the `Result<String, Error>`, not just the `String`.

Reply:
Isn't this only capturing the case where `find_task_rev` is `Ok(_)`. We are throwing the error away anyways. Could also just do  ` if let Ok(task_arn) = find_task_rev(ecs_client, &task_def_name, test_def).await {`.

Reply:
We're not trying to capture just `task_arn` though. We need to return a `Result<String, Error>`, so if we captured just the String on success we would just have to recreate the `Result<String, Error>` again. With this, if it is successful we just take that `Result` and return it. If it is _not_ successful, then we continue on and make the call to `create_task_definition`.

So we can change it that way, but it wouldn't be very efficient to just pull out the String and put it back into a new `Result`.